dynamic_instrumentation/src/instrument.rs

```diff
@@ -241,10 +241,12 @@ impl<'tcx> Visitor<'tcx> for InstrumentationAdder<'_, 'tcx> {
                     statement_index: location.statement_index + 1,
                     ..location
                 };
+
+                let source = remove_outer_deref(*p, self.tcx());
```

Comment:
`remove_outer_deref(*p, self.tcx())` is already computed above on line 238:

```rust
                    && place_ty(&remove_outer_deref(*p, self.tcx())).is_region_ptr() =>
```

Could you de-duplicate this?

Reply:
Maybe a `try_remove_outer_deref` would be useful here:

```rust
/// Try to strip the initital [`Deref`](ProjectionElem::Deref)
/// from a [`projection`](PlaceRef::projection) sequence.
pub fn try_remove_outer_deref<'tcx>(p: Place<'tcx>, tcx: TyCtxt<'tcx>) -> Option<Place<'tcx>> {
    // Remove outer deref if present
    match p.as_ref() {
        PlaceRef {
            local,
            projection: &[ref base @ .., ProjectionElem::Deref],
        } => Some(Place {
            local,
            projection: tcx.intern_place_elems(base),
        }),
        _ => None,
    }
}

/// Strip the initital [`Deref`](ProjectionElem::Deref)
/// from a [`projection`](PlaceRef::projection) sequence
/// if there is one.
pub fn remove_outer_deref<'tcx>(p: Place<'tcx>, tcx: TyCtxt<'tcx>) -> Place<'tcx> {
    try_remove_outer_deref(p, tcx).unwrap_or(p)
}
```

Reply:
Could do

```rust
            Rvalue::AddressOf(_, p)
                if let Some(source) = try_remove_outer_deref(*p, self.tcx())
                && place_ty(source).is_region_ptr() =>
            {
                let instrumentation_location = Location {
                    statement_index: location.statement_index + 1,
                    ..location
                };
                // Instrument which local's address is taken
                self.loc(instrumentation_location, copy_fn)
                    .arg_var(dest)
                    .source(source)
                    .dest(&dest)
                    .debug_mir(location)
                    .add_to(self);
            }
```

but that requires `#![feature(if_let_guard, let_chains)]`.  It might be too clunky without those, I think.

Reply:
How about the changes in https://github.com/immunant/c2rust/pull/573?

Reply:
I think https://github.com/immunant/c2rust/pull/573 is probably too big for the scope of this PR, but keep it in my mind for after merging this PR.